ipn/ipnlocal/network-lock.go

```diff
@@ -39,6 +39,36 @@ type tkaState struct {
 	storage   *tka.FS
 }
 
+// tkaFilterNetmapLocked checks the signatures on each node key, dropping
+// nodes from the netmap who's signature does not verify.
+func (b *LocalBackend) tkaFilterNetmapLocked(nm *netmap.NetworkMap) {
+	if !envknob.UseWIPCode() {
+		return // Feature-flag till network-lock is in Alpha.
+	}
+	if b.tka == nil {
+		return // TKA not enabled.
+	}
+
+	toDelete := make([]int, 0, len(nm.Peers))
+	for i, p := range nm.Peers {
+		if len(p.KeySignature) == 0 {
+			b.logf("Network lock is dropping peer %v(%v) due to missing signature", p.ID, p.StableID)
+			toDelete = append(toDelete, i)
+		} else {
+			if err := b.tka.authority.NodeKeyAuthorized(p.Key, p.KeySignature); err != nil {
+				b.logf("Network lock is dropping peer %v(%v) due to failed signature check: %v", p.ID, p.StableID, err)
+				toDelete = append(toDelete, i)
+			}
+		}
+	}
+
+	// nm.Peers is ordered, so deletion must be order-preserving.
```

Comment:
algorithmically problematic.

make toDelete a map with key idx, then create a new slice by iterating over the old peers slice, skipping each element that's in the toDelete map? (you can even reuse the memory I believe)

Reply:
FWIW, it is not too hard to permute a slice in-place to remove elements. It's linear time, but does not require any reallocation.

Example: https://github.com/creachadair/filter/blob/v0.1.0/filter.go#L19

It may not be worthwhile to do that, but I thought I'd point it out, since pushing into and out of a map is kind of heavyweight.

Reply:
Unfortunately I cant do the swap-to-back-and-truncate-slice trick as the ordering must remain intact.

Reply:
That's not what this code does. It preserves the order of the retained items.

Reply:
Oooo gotcha